doc/tikv/transaction-handling-newbie-perspective.md

```diff
@@ -0,0 +1,344 @@
+
+And the it returns whether the lock is successful. If not, it will also [return the lock to wait for](https://github.com/tikv/tikv/blob/bf716a111fde9fe8da56f8bd840c53d80c395525/src/storage/txn/process.rs#L447).
+
+#### Prewrite
+
+On TiKV side, the prewrite process happens [here in `process_write_impl`](https://github.com/tikv/tikv/blob/4a75902f266fbbc064f0c19a2a681cfe66511bc3/src/storage/txn/process.rs#L557).
+
+The first few lines of code (`if rows > FORWARD_MIN_MUTATIONS_NUM` part) is not covered by the [`TiKV Source Code Reading blogs`](https://pingcap.com/blog-cn/tikv-source-code-reading-12/). I guess it means:
+
+```
+if there's no "write" record in [mutations.minKey, mutation.maxKey] {
+	skip_constraint_check = true;
+  scan_mode = Some(ScanMode::Forward)
+}
+```
+
+As far as I understand, it just provides a optimized way of checking the "write" column, see [tikv#5846](https://github.com/tikv/tikv/pull/5846) for details.
+
+And no matter whether this branch is taken, we'll construct a `MvccTxn` , and then use it to do the prewrite job for each mutation the client sent to the TiKV server.
+
+The [`MvccTxn::prewrite`](https://github.com/tikv/tikv/blob/4a75902f266fbbc064f0c19a2a681cfe66511bc3/src/storage/mvcc/txn.rs#L563) function just do what the [Percolator](https://research.google/pubs/pub36726/) describes: check the `write` record in `[start_ts, ∞]` to find a newer write (this can be bypassed if `skip_constraint_check` is set, we can ignore this check safely in situations like import data). And then check whether the current key is locked at any timestamp. And finally use [`prewrite_key_value`](https://github.com/tikv/tikv/blob/bf716a111fde9fe8da56f8bd840c53d80c395525/src/storage/mvcc/txn.rs#L207) to lock the key and write the value in.
+
+##### Latches
+
+Just as I mentioned, there's no such things like "single row transaction" in TiKV, so we need another way to prevent the key's locking state changed by another transaction during `prewrite`.
+
+TiKV use [`Latches`](https://github.com/tikv/tikv/blob/bf716a111fde9fe8da56f8bd840c53d80c395525/src/storage/txn/latch.rs#L125) to archieve this, you can consider it as a Map from key('s hashcode) to mutexes. You can lock a key in the `Latches` to prevent it be used by other transactions.
+
+The latches is used in [`try_to_wake_up`](https://github.com/tikv/tikv/blob/bf716a111fde9fe8da56f8bd840c53d80c395525/src/storage/txn/scheduler.rs#L335) , this is called before each command is executed, it will lock all the latches the commands used.
+
+![prewrite-dfd](transaction-handling-newbie-perspective/prewrite-dfd.svg)
+
+#### PrewritePessimistic
+
+[`PrewritePessimistic`'s handling](https://github.com/tikv/tikv/blob/3a4a0c98f9efc2b409add8cb6ac9e8886bb5730c/src/storage/txn/process.rs#L624) is very similiar to `Prewrite`, except it:
+
+- doesn't need to read the write record for checking conflict
+- downgrade the pessimistic lock to optimistic lock after prewrite
+- needs to prevent deadlock
```

Comment:
Could you give an example of how deadlock can happen and why it isn't a problem for optimistic transactions please